python/lsst/daf/butler/datastores/fileDatastore.py

```diff
@@ -1647,60 +1633,85 @@ def emptyTrash(self, ignore_errors: bool = True) -> None:
+        # It will also automatically delete the relevant rows from the
+        # trash table and the records table.
+        with self.bridge.emptyTrash(self._table, record_class=StoredFileInfo,
+                                    record_column="path") as trash_data:
+            # Removing the artifacts themselves requires that the files are
+            # not also associated with refs that are not to be trashed.
+            # Therefore need to do a query with the file paths themselves
+            # and return all the refs associated with them. Can only delete
+            # a file if the refs to be trashed are the only refs associated
+            # with the file.
+            # This requires multiple copies of the trashed items
+            trashed, artifacts_to_keep = trash_data
+
+            if artifacts_to_keep is None:
+                # The bridge is not helping us so have to work it out
+                # ourselves. This is not going to be as efficient.
+                trashed = list(trashed)
+
+                # The instance check is for mypy since up to this point it
+                # does not know the type of info.
+                path_map = self._refs_associated_with_artifacts([info.path for _, info in trashed
+                                                                 if isinstance(info, StoredFileInfo)])
+
+                for ref, info in trashed:
+                    path_map[info.path].remove(ref.id)
+                    if not path_map[info.path]:
+                        del path_map[info.path]
+
+                artifacts_to_keep = set(path_map)
+
+            for ref, info in trashed:
+
+                if info is None:
+                    # Should not happen for this implementation but need
+                    # to keep mypy happy.
+                    raise RuntimeError(f"Internal logic error in emptyTrash with ref {ref}.")
+
+                if not isinstance(info, StoredFileInfo):
+                    # Mypy needs to know this is not the base class
+                    raise RuntimeError(f"Unexpectedly got info of class {type(info)}")
+
+                if ref.id is None:
+                    # Check to keep mypy happy.
+                    raise RuntimeError(f"Internal logic error in emptyTrash with ref {ref}/{info}")
+
+                # Only trashed refs still known to datastore will be returned.
+                location = info.file_location(self.locationFactory)
+
+                # If the file itself has been deleted there is nothing we
+                # can do about it. It is possible that trash has been run
+                # in parallel in another process or someone decided to delete
+                # the file. It is unlikely to come back and so we should still
+                # continue with the removal of the entry from the trash
+                # table.
+                if not self._artifact_exists(location):
```

Comment:
I think I might remove this check completely and move it to a catch below for FileNotFoundError. That way we save an existence check (which can add up with the S3 interface).